src/collection-view.js

```diff
@@ -321,9 +337,19 @@ Marionette.CollectionView = Marionette.View.extend({
     // remove and/or close it later
     this.children.add(view);
 
+    if (canTriggerAttach && this.triggerBeforeAttach) {
+      nestedViews = [view].concat(view._getNestedViews());
+      // Trigger before:attach after render to honor Region#show() lifecycle order.
+      view.once('render', _.partial(this._triggerMethodOnAll, nestedViews, 'before:attach'), this);
```

Comment:
Ahh, I thought about refactoring addEmptyView and _addChildView to eliminate duplicated code, but decided not to take that on for pull request size and without more research (why are they duplicated in the first place).  I think that would be a good idea.

I'm glad you asked about `once` render.  The very next line is `this.renderChildView`:

```js
  renderChildView: function(view, index) {
    view.render();
    this.attachHtml(this, view, index);
    return view;
  },
```

`renderChildView` is part of the public API and a suggested override for custom child view rendering.  I noticed a pattern within collection-view.js that override-friendly public API methods do not trigger lifecycle events, which seemed reasonable because then the developer would be responsible for lifecycle.  The issue here that _before:attach_ comes after _render_ but before the html is attached to the DOM.  The only other way to code this would be:

```js
  renderChildView: function(view, index) {
    view.render();
    Marionette.triggerMethodOn(view, 'before:attach');
    this.attachHtml(this, view, index);
    return view;
  },
```

Triggering _before:attach_ seemed like too much responsibility for the overridable public method in my gut, but also based on the pattern I noticed elsewhere.